client/executor/wasmtime/src/imports.rs

```diff
@@ -0,0 +1,327 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::state_holder::StateHolder;
+use sc_executor_common::error::WasmError;
+use sp_wasm_interface::{Function, Value, ValueType};
+use std::any::Any;
+use std::rc::Rc;
+use wasmtime::{
+	Callable, Extern, ExternType, Func, FuncType, ImportType, Limits, Memory, MemoryType, Module,
+	Trap, Val,
+};
+
+pub struct Imports {
+	/// Contains the index into `externs` where the memory import is stored if any. `None` if there
+	/// is none.
+	pub memory_import_index: Option<usize>,
+	pub externs: Vec<Extern>,
+}
+
+/// Goes over all imports of a module and prepares a vector of `Extern`s that can be used for
+/// instantiation of the module. Returns an error if there are imports that cannot be satisfied.
+pub fn resolve_imports(
+	state_holder: &StateHolder,
+	module: &Module,
+	host_functions: &[&'static dyn Function],
+	heap_pages: u32,
+	allow_missing_func_imports: bool,
+) -> Result<Imports, WasmError> {
+	let mut externs = vec![];
+	let mut memory_import_index = None;
+	for import_ty in module.imports() {
+		if import_ty.module() != "env" {
+			return Err(WasmError::Other(format!(
+				"host doesn't provide any imports from non-env module: {}:{}",
+				import_ty.module(),
+				import_ty.name()
+			)));
+		}
+
+		let resolved = match import_ty.name() {
+			"memory" => {
+				memory_import_index = Some(externs.len());
```

Comment:
Should we error when a second `memory` instance is requested?

Reply:
I am not sure!

First, it won't be possible atm, since if the module has 2 memory imports then it is a not valid module. Second, as soon as multiple memories are supported then it is OK (although useless) to import the same memory twice.